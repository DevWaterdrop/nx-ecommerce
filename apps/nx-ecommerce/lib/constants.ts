export const LOVELY_ERROR = 'I HATE TYPESCRIPT+STRAPI+GRAPHQL';
export const MAXIMUM_IN_CART = 16;

export enum PRICE_FILTER {
  ZERO_TWENTY = 'zt',
  TWENTY_FORTY = 'tf',
  FORTY_EIGHTY = 'fe',
  EIGHTY_PLUS = 'ep',
}

export enum SORT_FILTER {
  PRICE_LOW_HIGH = 'plh',
  PRICE_HIGH_LOW = 'phl',
  NAME = 'name',
}

export const PRICES = [
  { value: PRICE_FILTER.ZERO_TWENTY, label: '0.00$ - 19.99$' },
  { value: PRICE_FILTER.TWENTY_FORTY, label: '20.00$ - 39.99$' },
  { value: PRICE_FILTER.FORTY_EIGHTY, label: '40.00$ - 79.99$' },
  { value: PRICE_FILTER.EIGHTY_PLUS, label: '80.00$+' },
];

export const SORT = [
  { value: SORT_FILTER.PRICE_LOW_HIGH, label: 'Price: low to high' },
  { value: SORT_FILTER.PRICE_HIGH_LOW, label: 'Price: high to low' },
  { value: SORT_FILTER.NAME, label: 'Name' },
];
