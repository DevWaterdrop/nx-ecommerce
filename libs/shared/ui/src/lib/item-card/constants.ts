import { ItemCardProps } from './item-card';
import flowerJPEG from '../../props/flower.jpeg';
import roomJPEG from '../../props/room.jpeg';

const ITEM: Required<ItemCardProps['item']> = {
  price: 65,
  name: 'Lovely flower',
  id: '#',
  images: [flowerJPEG, roomJPEG],
  smallDescription: 'For you, for me, for us',
  tag: 'new',
};

export const DEFAULT_PROPS: ItemCardProps = {
  item: ITEM,
  size: 'base',
  handleClick: () => console.log('clicked'),
};
