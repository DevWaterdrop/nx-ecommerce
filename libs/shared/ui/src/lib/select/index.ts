import { Select, SelectProps } from './select';

export { Select };

export type { SelectProps };
