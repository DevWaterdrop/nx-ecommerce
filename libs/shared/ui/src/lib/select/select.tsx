import { useState, useRef } from 'react';
import { useClickOutside } from '@nx-ecommerce/shared/utils/use-click-outside';
import { Button } from './local-components/button';
import type { Props as DropdownProps } from './local-components/dropdown';
import dynamic from 'next/dynamic';

export type Item = { label: string; value: string };
export type ItemMap = Map<Item['value'], string>;
type SelectedState<T extends boolean> = T extends true ? ItemMap : Item;

/* eslint-disable-next-line */
export interface SelectProps {
  items: Item[];
  onChange: (item?: string | string[]) => void;
  initial?: string | string[];
  placeholder?: string;
  multiple?: boolean;
}

export function createInitial(
  initial: SelectProps['initial'],
  items: SelectProps['items'],
  multiple: SelectProps['multiple']
) {
  if (!initial) return multiple ? new Map() : null;

  if (initial instanceof Array || multiple) {
    const map = items.reduce<[string, string][]>((prev, curr) => {
      if (initial.includes(curr.value)) {
        return [...prev, [curr.value, curr.label]];
      }

      return prev;
    }, []);

    return new Map(map);
  }

  return items.find(({ value }) => value === initial) || null;
}

function createSelected(prev: Item | ItemMap | null, item: Item) {
  const isMap = prev instanceof Map;

  if (isMap) {
    const copy = new Map(prev);
    const existInSelected = copy.has(item.value);

    if (existInSelected) {
      copy.delete(item.value);
    } else {
      copy.set(item.value, item.label);
    }

    return copy;
  }

  if (prev) {
    const same = prev.value === item.value;
    if (same) return null;
  }

  return item;
}

const LazyDropdown = dynamic<DropdownProps>(() =>
  import('./local-components/dropdown').then(({ Dropdown }) => Dropdown)
);

export const Select: React.FC<SelectProps> = (props) => {
  const {
    items,
    placeholder = 'Select',
    multiple = false,
    onChange,
    initial,
  } = props;

  const [selected, setSelected] = useState<SelectedState<
    typeof multiple
  > | null>(createInitial(initial, items, multiple));

  const [open, setOpen] = useState(false);

  const containerRef = useRef<HTMLDivElement>(null);

  useClickOutside(containerRef, () => setOpen(false));

  const handleItemChange = (item: Item) => {
    setSelected((prev) => {
      const newItem = createSelected(prev, item);

      onChange(newItem instanceof Map ? [...newItem.keys()] : newItem?.value);
      return newItem;
    });
  };

  return (
    <div className="rotate relative" ref={containerRef}>
      <Button
        open={open}
        placeholder={placeholder}
        setOpen={setOpen}
        selected={selected}
      />
      {open && (
        <LazyDropdown
          handleItemChange={handleItemChange}
          items={items}
          selected={selected}
        />
      )}
    </div>
  );
};
