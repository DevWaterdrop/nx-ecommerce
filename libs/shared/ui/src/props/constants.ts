import { ButtonProps } from '../lib/button';
import { CartItemProps } from '../lib/cart-item';
import { CartSidebarProps } from '../lib/cart-sidebar';
import { CategoryCardProps } from '../lib/category-card';
import { IconProps } from '../lib/icon/icon';
import { ItemCardProps } from '../lib/item-card';
import { NavButtonProps } from '../lib/nav-button';
import { OrdersTableProps } from '../lib/orders-table';
import { ProductGalleryProps } from '../lib/product-gallery';
import { ProductInformationProps } from '../lib/product-information';
import { ProductSidebarProps } from '../lib/product-sidebar';
import { SearchbarProps } from '../lib/searchbar';
import { SearchbarItemProps } from '../lib/searchbar-item';
import { SelectProps } from '../lib/select';
import flowerJPEG from './flower.jpeg';
import roomJPEG from './room.jpeg';

export const ITEM: ItemCardProps['item'] = {
  slug: 'lovely-flower',
  price: 65,
  name: 'Lovely flower',
  images: {
    data: [
      {
        attributes: {
          url: flowerJPEG,
        },
      },
      {
        attributes: {
          url: roomJPEG,
        },
      },
    ],
  },
  smallDescription: 'For you, for me, for us',
  description: 'Lovely description',
};

export const ITEM_SECOND: ItemCardProps['item'] = {
  slug: 'roomy',
  price: 50,
  name: 'Roomy',
  images: {
    data: [
      {
        attributes: {
          url: flowerJPEG,
        },
      },
      {
        attributes: {
          url: roomJPEG,
        },
      },
    ],
  },
  smallDescription: 'Hmmm, nice vase',
  description: 'Vaaaase',
};

export const ITEM_CARD_DEFAULT_PROPS: ItemCardProps = {
  item: ITEM,
  size: 'base',
  handleCartClick: () => console.log('cart clicked'),
  handleAmountCartClick: () => console.log('amount cart clicked'),
  handleFavoriteClick: () => console.log('favorite clicked'),
};

export const CATEGORY_CARD_DEFAULT_PROPS: CategoryCardProps = {
  category: {
    smallDescription:
      'Lorem ipsum dolor sit amet, consectetur adipisicing elit.',
    image: {
      data: {
        attributes: {
          url: roomJPEG,
        },
      },
    },
    slug: '1',
  },
};

export const NAV_BUTTON_DEFAULT_PROPS: NavButtonProps = {
  type: 'heart',
};

export const ICON_DEFAULT_PROPS: IconProps = {
  icon: 'cart',
};

export const BUTTON_DEFAULT_PROPS: ButtonProps = {
  children: 'hello',
  tag: 'button',
  size: 'base',
  type: 'primary',
  round: 'full',
};

export const SEARCHBAR_ITEM: SearchbarProps['items'][number] = {
  id: '111',
  attributes: {
    slug: 'lovely-flower',
    smallDescription: 'For you, for me, for us',
    name: 'Lovely flower',
    images: {
      data: [
        {
          attributes: {
            url: flowerJPEG,
          },
        },
      ],
    },
  },
};

export const SEARCHBAR_ITEM_DEFAULT_PROPS: SearchbarItemProps = {
  item: {
    slug: 'lovely-flower',
    smallDescription: 'For you, for me, for us',
    name: 'Lovely flower',
    images: {
      data: [
        {
          attributes: {
            url: flowerJPEG,
          },
        },
      ],
    },
  },
  handleClick: () => void 1,
};

export const SELECT_ITEM = {
  label: 'Label1',
  value: 'value1',
};

export const SELECT_DEFAULT_PROPS: Required<
  Omit<SelectProps, 'placeholder' | 'multiple'>
> = {
  items: [SELECT_ITEM],
  initial: ['value1'],
  onChange: () => console.log('changed'),
};

export const PRODUCT: ProductSidebarProps['product'] = {
  slug: 'lovely-flower',
  price: 65,
  name: 'Lovely flower',
  images: {
    data: [
      {
        attributes: {
          url: flowerJPEG,
        },
      },
      {
        attributes: {
          url: roomJPEG,
        },
      },
    ],
  },
  smallDescription: 'For you, for me, for us',
  description: 'Lovely description',
  information: [
    {
      title: 'Title1',
      content: 'content1',
    },
  ],
};

export const PRODUCT_SIDEBAR_DEFAULT_PROPS: ProductSidebarProps = {
  product: PRODUCT,
  handleCartClick: () => {
    console.log('cart');
  },
  handleRemoveCartClick: () => {
    console.log('remove cart');
  },
  handleFavoriteClick: () => {
    console.log('favorite');
  },
  cartValue: new Map([]),
  favoritesValue: new Set('lovely-flower'),
};

export const PRODUCT_INFORMATION_DEFAULT_PROPS: ProductInformationProps = {
  information: [...Array(10).keys()].map((_value, index) => ({
    title: `Title${index + 1}`,
    content: `content${index + 1}`,
  })),
};

export const PRODUCT_GALLERY_DEFAULT_PROPS: ProductGalleryProps = {
  images: [
    {
      attributes: {
        url: flowerJPEG,
      },
    },
    {
      attributes: {
        url: roomJPEG,
      },
    },
  ],
};

const PRODUCT_ORDER_TABLE: OrdersTableProps['orders'][number]['items'][number]['product'] =
  {
    name: 'Flower',
    slug: 'slug-flower',
    description: '',
    smallDescription: '',
    price: 50,
  };

export const ORDERS_TABLE_DEFAULT_PROPS: OrdersTableProps = {
  orders: [
    {
      id: 'id-1',
      amount: 50,
      items: [{ id: '1', amount: 1, product: PRODUCT_ORDER_TABLE }],
    },
    {
      id: 'id-2',
      amount: 100,
      items: [{ id: '2', amount: 2, product: PRODUCT_ORDER_TABLE }],
    },
  ],
};

export const CART_SIDEBAR_DEFAULT_PROPS: CartSidebarProps = {
  cartTotal: 100,
  isUser: true,
  handleCheckoutClick: async () => {
    console.log('checkout');
  },
  isLoading: false,
};

export const CART_ITEM_DEFAULT_PROPS: CartItemProps = {
  inCartAmount: 1,
  item: PRODUCT,
  handleCartMinusClick: () => {
    console.log('cart minus');
  },
  handleCartPlusClick: () => {
    console.log('cart plus');
  },
  handleRemoveFromCart: () => {
    console.log('cart remove');
  },
  toggleFavorite: () => {
    console.log('favorite');
  },
};
