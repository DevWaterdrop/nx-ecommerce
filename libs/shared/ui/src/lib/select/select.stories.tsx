import { SELECT_DEFAULT_PROPS } from '../../props/constants';
import { Story, Meta } from '@storybook/react';
import { Select, SelectProps } from './select';

export default {
  component: Select,
  title: 'Components/Select',
} as Meta;

const Template: Story<SelectProps> = (args) => <Select {...args} />;

export const Primary = Template.bind({});
Primary.args = { ...SELECT_DEFAULT_PROPS };
