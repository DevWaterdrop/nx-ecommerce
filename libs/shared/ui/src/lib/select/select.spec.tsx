import { render, RenderResult } from '@testing-library/react';
import '@testing-library/jest-dom';

import { SELECT_DEFAULT_PROPS } from '../../props/constants';
import { createInitial, Select } from './select';

describe('Select', () => {
  let renderResult: RenderResult;

  beforeEach(() => {
    renderResult = render(<Select {...SELECT_DEFAULT_PROPS} />);
  });

  it('should render successfully', () => {
    const { baseElement } = renderResult;
    expect(baseElement).toBeTruthy();
  });

  it('successfully create initial state', () => {
    const { initial, items } = SELECT_DEFAULT_PROPS;

    const TEST_CASE_1 = [undefined, items, false] as const;
    expect(createInitial(...TEST_CASE_1)).toBe(null);

    const TEST_CASE_2 = [undefined, items, true] as const;
    expect(createInitial(...TEST_CASE_2)).toBeInstanceOf(Map);

    const TEST_CASE_3 = [initial, items, true] as const;
    const testCase3Map = createInitial(...TEST_CASE_3);
    expect(testCase3Map).toBeInstanceOf(Map);
    expect((testCase3Map as Map<string, unknown>).has(initial[0])).toBe(true);

    const TEST_CASE_4 = [initial[0], items, false] as const;
    expect(createInitial(...TEST_CASE_4)).toBe(items[0]);

    const TEST_CASE_5 = [initial[0], items, true] as const;
    expect(createInitial(...TEST_CASE_5)).toBeInstanceOf(Map);

    const TEST_CASE_6 = ['wrong key', items, false] as const;
    expect(createInitial(...TEST_CASE_6)).toBe(null);
  });
});
