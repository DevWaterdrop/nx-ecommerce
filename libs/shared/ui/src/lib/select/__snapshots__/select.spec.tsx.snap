// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render successfully 1`] = `
<div
  className="rotate relative"
>
  <button
    className="py-2.5 px-4 flex items-end gap-2 bg-gray-100 text-sm font-medium rounded-full border border-transparent sm:(px-6) border-black"
    onClick={[Function]}
    type="button"
  >
    Select
    <div
      aria-hidden="true"
      className="h-4.5 w-4.5 relative transition-transform transform-gpu"
    >
      Next image stub
    </div>
  </button>
</div>
`;
